Always select non link-local network interface, fixes #11961.

paralin opened the pull request:
Fixes #11961. The API server will refuse to start if a link-local IP address is selected, however, ChooseNetworkInterface can still select link-local network routes. This causes a startup failure on platforms like DigitalOcean that use a link-local route to fetch metadata.

paralin commented:
I signed it!

That bot is so fast it posted 2 replies!

thockin commented:
ok to test

thockin commented:
Code looks good.  2 remarks:

1: something funky in your 2nd commit - can you please squash to a single commit and see that it produces the minimal change
2: please add a test case in `TestChooseHostInterfaceFromRoute`

paralin committed:
always select non link-local interface, fixes #11961

The API server will refuse to start if a link-local IP address is selected, however, ChooseNetworkInterface can still select link-local network routes. This causes a startup failure on platforms like DigitalOcean that use a link-local route to fetch metadata.

Signed-off-by: Christian Stewart <christian@paral.in>

paralin commented:
@thockin Cleaned up the commits, give me a few minutes to implement the test.

paralin commented:
Okay @thockin I took a sample of /proc/net/route from a real DigitalOcean instance and wrote the test around that. It appears that it will select the correct IP address now.

paralin commented:
Also the test failure on the last commit was due to a Shippable authentication problem with GitHub...

paralin commented:
@thockin Fixed the case in the variable, fixed the /proc/net/route entry, and corrected the Addrs function in the test.

thockin commented on pkg/util/util_test.go:
```diff
@@ -499,6 +505,19 @@ func (_ validNetworkInterface) Addrs(intf *net.Interface) ([]net.Addr, error) {
+}
+func (_ validNetworkInterfaceWithoutLinkLocal) Addrs(intf *net.Interface) ([]net.Addr, error) {
+	var ifat []net.Addr
+	ifat = []net.Addr{addrStruct{val: "45.55.47.146/19"}, addrStruct{val: "169.254.162.166/16"}}
```
shouldn't the link-local come first, to prove it does not get selected?

paralin replied:
Whoops.

paralin commented:
@thockin For whatever reason if I invert the two in the test I'm getting an error that the code cannot select an IP address. Very very weird. I'll keep looking into it.

paralin commented:
Okay the issue is in `getIPFromInterface` it just gets the first IP address (I think). Working on this.

paralin commented:
Okay, the solution is to move the check into `getFinalIP`. Pushing the fix.

paralin commented:
@thockin Fixes complete, the tests pass on my end. I amended the comment on `getFinalIP` since it will now skip IPs that are loopback (as before) OR link-local. This doesn't seem to affect anything but it's worth noting.

paralin commented:
This apparently affected the DNS backend tests... Will investigate more tomorrow.

thockin commented on pkg/util/util_test.go:
```diff
@@ -499,6 +505,19 @@ func (_ validNetworkInterface) Addrs(intf *net.Interface) ([]net.Addr, error) {
 	return ifat, nil
 }
 
+type validNetworkInterfaceWithoutLinkLocal struct {
```
one more nit:  s/Without/With ?

paralin replied:
Maybe. It checks that it *doesn't* select link local, but the test really is about if the network interface is selected properly with a link-local in the pool...

e2e is broken though so I think that's the bigger issue.

paralin commented:
Interesting, now it's working in the e2e tests. I think something else might have fixed them in the commit history. I'd say this is ready for merge now, @thockin

paralin commented:
Changed the test from WithoutLinkLocal to WithLinkLocal

thockin commented:
lgtm
